util/upload-cookbook.py

```diff
@@ -0,0 +1,241 @@
+                aws_session_token=aws_credentials.get("SessionToken"),
+            )
+
+        except ClientError:
+            print("Warning: non authorized in region '{0}', skipping".format(credential_region))
+            pass
+    else:
+        s3 = boto3.client("s3")
+    return s3
+
+
+def _get_bucket_name(args, region):
+    return region + "-aws-parallelcluster" if not args.bucket else args.bucket
+
+
+def main(args):
+
+    for _region in args.regions:
+        s3 = _create_s3_client(_region)
+        bucket1_name = _get_bucket_name(args, _region)
+
+        print("Listing cookbook for region ({0})".format(_region))
+        _aws_ls(s3, _region, bucket1_name, _COOKBOOKS_DIR + "/" + args.full_name + ".tgz")
+
+    if len(_ls_error_array) > 0 and not args.override:
+        print("We know the cookbook archives are already there, in this round we need to upload the .date files!")
+        print("Failed to push cookbook, already present for region " + " ".join(_ls_error_array))
+        exit(1)
+    elif len(_ls_error_array) > 0 and args.override:
+        print("Some or all of the cookbook archives are already there but OVERRIDE=true")
+
+    for _region in args.regions:
+        s3 = _create_s3_client(_region)
+        bucket1_name = _get_bucket_name(args, _region)
+
+        if args.override:
+            print("Backup cookbook for region :{0}".format(_region))
+            _aws_bck(s3, args, _region, bucket1_name, args.full_name + ".tgz")
+            _aws_bck(s3, args, _region, bucket1_name, args.full_name + ".md5")
+            _aws_bck(s3, args, _region, bucket1_name, args.full_name + ".tgz.date")
+
+        print("Pushing cookbook for region: {0}".format(_region))
+        _aws_cp(s3, args, _region, bucket1_name, _COOKBOOKS_DIR, args.full_name + ".tgz")
+        _aws_cp(s3, args, _region, bucket1_name, _COOKBOOKS_DIR, args.full_name + ".md5")
+
+        if not args.dryrun:
+            # Stores LastModified info into .tgz.date file and uploads it back to bucket
+            with (open(args.full_name + ".tgz.date", "w+")) as f:
+                response = s3.head_object(Bucket=bucket1_name, Key=_COOKBOOKS_DIR + "/" + args.full_name + ".tgz")
+                f.write(response.get("LastModified").strftime("%Y-%m-%d_%H-%M-%S"))
+
+            _aws_cp(s3, args, _region, bucket1_name, _COOKBOOKS_DIR, args.full_name + ".tgz.date")
+        else:
+            print(
+                "File {0}.{1} not stored to bucket {2} due to dryrun mode".format(
+                    args.full_name, ".tgz.date", bucket1_name
+                )
+            )
+
+    if len(_bck_error_array) > 0:
+        print("Failed to backup cookbook for region ({0})".format(" ".join(_bck_error_array)))
+
+    if len(_cp_error_array) > 0:
+        print("Failed to push cookbook for region ({0})".format(" ".join(_cp_error_array)))
+        exit(1)
+
+
+if __name__ == "__main__":
+    parser = argparse.ArgumentParser(description="Uploads cookbook to S3")
```

Comment:
I'd move all these parsing in a _get_parser function, see for example:
https://github.com/aws/aws-parallelcluster/blob/v2.4.1/cli/awsbatch/awsbstat.py#L38

Reply:
Then I'd move the _get_parser into the main function

Reply:
ok